src/v/cluster/cluster_discovery.cc

```diff
@@ -142,8 +202,259 @@ ss::future<bool> cluster_discovery::dispatch_node_uuid_registration_to_seeds(
+ss::future<cluster_bootstrap_info_reply>
+cluster_discovery::request_cluster_bootstrap_info_single(
+  const net::unresolved_address addr) const {
+    vlog(clusterlog.info, "Requesting cluster bootstrap info from {}", addr);
+    _as.check();
+    for (auto repeat_jitter = simple_time_jitter<model::timeout_clock>(1s);;) {
+        result<cluster_bootstrap_info_reply> reply_result(
+          std::errc::connection_refused);
+        try {
+            reply_result = co_await do_with_client_one_shot<
+              cluster_bootstrap_client_protocol>(
+              addr,
+              config::node().rpc_server_tls(),
+              2s,
+              [](cluster_bootstrap_client_protocol c) {
+                  return c
+                    .cluster_bootstrap_info(
+                      cluster_bootstrap_info_request{},
+                      rpc::client_opts(rpc::clock_type::now() + 2s))
+                    .then(&rpc::get_ctx_data<cluster_bootstrap_info_reply>);
+              });
+            if (reply_result) {
+                vlog(
+                  clusterlog.info,
+                  "Obtained cluster bootstrap info from {}",
+                  addr);
+                vlog(clusterlog.debug, "{}", reply_result.value());
+                co_return std::move(reply_result.value());
+            }
+            vlog(
+              clusterlog.debug,
+              "Cluster bootstrap info failed with {}",
+              reply_result.error());
+        } catch (...) {
+            vlog(
+              clusterlog.warn,
+              "Error requesting cluster bootstrap info from {}, retrying. {}",
+              addr,
+              std::current_exception());
+        }
+        co_await ss::sleep_abortable(repeat_jitter.next_duration(), _as);
+        vlog(clusterlog.trace, "Retrying cluster bootstrap info from {}", addr);
+    }
+}
+
+ss::future<> cluster_discovery::fetch_cluster_bootstrap_info() {
+    if (_cluster_bootstrap_info_fetched) {
+        co_return;
+    }
+    const auto units = co_await _fetch_cluster_bootstrap_mx.get_units(_as);
+    if (_cluster_bootstrap_info_fetched) {
+        co_return;
+    }
+
+    const net::unresolved_address& self_addr
+      = config::node().advertised_rpc_api();
+    const std::vector<config::seed_server>& self_seed_servers
+      = config::node().seed_servers();
+
+    std::vector<
+      std::pair<net::unresolved_address, cluster_bootstrap_info_reply>>
+      peers;
+    peers.reserve(self_seed_servers.size());
+    for (const config::seed_server& seed_server : self_seed_servers) {
+        // do not call oneself
+        if (seed_server.addr == self_addr) {
+            continue;
+        }
+        peers.emplace_back(seed_server.addr, cluster_bootstrap_info_reply{});
+    }
+    co_await ss::parallel_for_each(peers, [this](auto& peer) -> ss::future<> {
+        peer.second = co_await request_cluster_bootstrap_info_single(
+          peer.first);
+        co_return;
+    });
```

Comment:
nit: it seems we have spread some of the validations across different calls in `cluster_discovery`. I think it would make reading through this much easier if all validations happened up front. Is that feasible, such that by the end of this function we know:
- whether we should be founding a cluster (by the absence of a cluster UUID)
- if there are cluster UUIDs, that they all match
- that all seeds' configurations match if we _are_ a founder
- that all seeds' versions match if we _are_ a founder

Then we can remove some code from `get_seed_brokers_for_bootstrap` and `remote_cluster_uuid_exists`. At least for me, as I walk through the code, it's really hard to keep track of what has been validated and what hasn't, and I think doing a bunch of checks up front would help. It also helps us fail fast.